src/operator/semijoin.rs

```diff
@@ -0,0 +1,609 @@
+    }
+
+    fn semijoin_stream_core_inner<F, Keys, Out>(
+        &self,
+        keys: &Stream<Circuit<S>, Keys>,
+        semijoin_func: F,
+    ) -> Stream<Circuit<S>, Out>
+    where
+        F: Fn(&Pairs::Key, &Pairs::Val) -> Out::Key + 'static,
+        // TODO: Associated type bounds (rust/#52662) really simplify things
+        // TODO: Allow non-unit timestamps
+        Pairs: BatchReader<Time = ()> + Data,
+        Pairs::Key: Clone + Ord,
+        Pairs::Val: Clone,
+        Pairs::R: HasZero + MulByRef,
+        Keys: BatchReader<Key = Pairs::Key, Val = (), Time = (), R = Pairs::R> + Data,
+        // TODO: Should this be `IndexedZSet<Key = Pairs::Key, Val = Pairs::Val>`?
+        Out: ZSet<R = Pairs::R> + 'static,
+    {
+        self.circuit()
+            .add_binary_operator(SemiJoinStream::new(semijoin_func), self, keys)
+    }
+}
+
+/// Semijoin two streams of batches, see [`Stream::semijoin_stream`]
+pub struct SemiJoinStreamMonotonic<Pairs, Keys, Out> {
+    _types: PhantomData<(Pairs, Keys, Out)>,
+}
+
+impl<Pairs, Keys, Out> SemiJoinStreamMonotonic<Pairs, Keys, Out> {
+    pub const fn new() -> Self {
+        Self {
+            _types: PhantomData,
+        }
+    }
+}
+
+impl<Pairs, Keys, Out> Operator for SemiJoinStreamMonotonic<Pairs, Keys, Out>
+where
+    Pairs: 'static,
+    Keys: 'static,
+    Out: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::Borrowed("SemiJoinStreamMonotonic")
+    }
+
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        true
+    }
+}
+
+impl<Pairs, Keys, Out> BinaryOperator<Pairs, Keys, Out>
+    for SemiJoinStreamMonotonic<Pairs, Keys, Out>
+where
+    Pairs: BatchReader<Time = ()> + 'static,
+    Pairs::Key: Clone + Ord,
+    Pairs::Val: Clone,
+    Pairs::R: HasZero + MulByRef,
+    Keys: BatchReader<Key = Pairs::Key, Val = (), Time = (), R = Pairs::R> + 'static,
+    Out: ZSet<Key = (Pairs::Key, Pairs::Val), R = Pairs::R> + 'static,
+{
+    fn eval(&mut self, pairs: &Pairs, keys: &Keys) -> Out {
+        let mut pair_cursor = pairs.cursor();
+        let mut key_cursor = keys.cursor();
+
+        // Choose capacity heuristically.
+        let mut builder = Out::Builder::with_capacity((), min(pairs.len(), keys.len()));
+
+        // While both keys are valid
+        // TODO: Is there a better way to iterate here? `keys_cursor` is the
+        //       thing really driving this, so can we just use it as the
+        //       source of iteration to do the least work possible?
+        while key_cursor.key_valid() && pair_cursor.key_valid() {
+            match key_cursor.key().cmp(pair_cursor.key()) {
+                // Match up both the cursors
+                Ordering::Less => key_cursor.seek_key(pair_cursor.key()),
+                Ordering::Greater => pair_cursor.seek_key(key_cursor.key()),
+
+                Ordering::Equal => {
+                    // TODO: Can the value of `()` ever be invalid? Do we need an `if
```

Comment:
Nope, there's always precisely one value when value type is `()`. But as per my comment above, I'm not sure we need to constrain the second input to semijoin to `Val = ()`.